### PR TITLE
Extract error handling and transformation logic

### PR DESCRIPTION
**Pull request by kfrz**
## Description of change
This PR attempts to address some of the complexities in `ApplicationController`, related to rescuing and coercing error classes. The purpose of transforming an error class is so that `vets-api` can exercise finer-grained control over the specific responses the client consumes. 

E.g., if the `ApplicationController` rescues an `ActionController::ParameterMissing` exception, we expect it to transform into a `Common::Exceptions::ParameterMissing` object, which inherits from `Common::Exceptions::BaseError`. 

`Common::Exceptions::BaseError` is used throughout `vets-api` as a parent class which provides support for custom messaging via `i18n` keys in `exceptions.en.yml`. This key-value mapping allows developers to normalize the messaging sent when disparate backend service errors are encountered, as the consistency in quality of "original" messages is low.

### Notable changes
- Use default Raven configuration for skipping exception types
- Move `SentryLogging` context logic into  module
- Move error transformation/coercion logic into a separate class 
- Add a common ErrorHandler to handle error coercion logic

## Testing done
- [ ] Spec suite ran via Jenkins CI
- [ ] Unit tests written for `ErrorHandler`

## Testing planned
- Will test on staging when PR is approved

#### Unique to this PR
- [ ] Error coercion logic is encapsulated into its own class
- [ ] Sentry logging and tagging is extracted into `SentryLogging` module

#### Applies to all PRs

- [x] Appropriate logging
- [x] ~Swagger docs have been updated, if applicable~
- [x] Provide link to originating GitHub issue, or connected to it via ZenHub
- [x] Does not contain any sensitive information (i.e. PII/credentials/internal URLs/etc., in logging, hardcoded, or in specs)
- [x] ~Provide which alerts would indicate a problem with this functionality (if applicable)~

### CODE REVIEW

File: app/controllers/concerns/ignore_not_found.rb [hunk 1]
```diff
@@ -2,6 +2,6 @@
 
 module IgnoreNotFound
   def skip_sentry_exception_types
-    ApplicationController::SKIP_SENTRY_EXCEPTION_TYPES + [Common::Exceptions::RecordNotFound]
+    SentryLogging::SKIP_SENTRY_EXCEPTION_TYPES + [Common::Exceptions::RecordNotFound]
```
Comment (johnpaulashenfelter) [inline app/controllers/concerns/ignore_not_found.rb]: Sentry has a built-in way to do this: https://docs.sentry.io/clients/ruby/config/ -- maybe it didn't when all this code was originally written? Seems like it would be much easier to move this to Sentry config

File: lib/sentry_logging.rb [hunk 2]
```diff
@@ -1,6 +1,74 @@
...
+  # rubocop:disable Metrics/CyclomaticComplexity
+  # rubocop:disable Metrics/PerceivedComplexity
+  def log_error(error)
+    if skip_sentry_exception_types.include?(error.class)
+      Rails.logger.error error.message.to_s, backtrace: error.backtrace
+    else
+      extra = error.respond_to?(:errors) ? { errors: error.errors.map(&:to_hash) } : {}
+      if error.is_a?(Common::Exceptions::BackendServiceException)
+        # Add additional user specific context to the logs
+        if @current_user.present?
```
Comment (johnpaulashenfelter) [inline lib/sentry_logging.rb]: Should this be a responsibility of the `Common::Exceptions::BackendServiceException`?

File: lib/sentry_logging.rb [hunk 3]
```diff
@@ -1,6 +1,74 @@
...
+    Common::Exceptions::RoutingError,
+    Common::Exceptions::Forbidden,
+    Breakers::OutageException
+  ].freeze
+
+  # rubocop:disable Metrics/CyclomaticComplexity
+  # rubocop:disable Metrics/PerceivedComplexity
+  def log_error(error)
+    if skip_sentry_exception_types.include?(error.class)
+      Rails.logger.error error.message.to_s, backtrace: error.backtrace
```
Comment (johnpaulashenfelter) [inline lib/sentry_logging.rb]: Doesn't this happen automatically in the log? This is basically "log it if we don't sentry it"

File: app/controllers/concerns/ignore_not_found.rb (hunk 1 above)
Comment (LindseySaari) [reply app/controllers/concerns/ignore_not_found.rb]: We could possibly move the exception types defined in the SKIP_SENTRY_EXCEPTION_TYPES constant to the config, but it looks like the method gets custom defined in other places as well

File: lib/sentry_logging.rb (hunk 3 above)
Comment (jholton) [reply lib/sentry_logging.rb]: The error passed to this method has been `rescue`d, so it wouldn't otherwise appear in the `Rails` log. Right?

File: lib/sentry_logging.rb (hunk 2 above)
Comment (jholton) [reply lib/sentry_logging.rb]: `Common::Exceptions::BackendServiceException` is raised in a number of places, and is a parent of class of a few of other error classes , many/most of which don't have access to the `@current_user` object.
That all raises the following question for me:  Should this code exist here, or in a library explicitly intended for use in `controller`s (or just refactored in the `ApplicationController`?  This library is used/included in controllers, AR models, Redis models, service classes, Sidekiq workers, etc, and most/all of this explicitly deals with `controller` concerns (request objects, current_user objects, etc)

Comment (kfrz) [reply lib/sentry_logging.rb]: Yes, I think so - the logic here is one of the most confusing aspects of errors across the platform and I think extracting the context building to the caller as Sentry doesn't (or shouldn't) care/know about any context specifics it just logs hashes of data

Comment (kfrz) [reply lib/sentry_logging.rb]: I think if we were to move this responsibility into the `::BackendServiceException` class it would be obfuscated, we may want more explicitness around the adding of user context info to Sentry

File: lib/sentry_logging.rb (hunk 3 above)
Comment (LindseySaari) [reply lib/sentry_logging.rb]: Really specific comment, but we can probably use the rails_logger method on line 15 instead of Rails.logger.error

File: app/controllers/concerns/ignore_not_found.rb (hunk 1 above)
Comment (kfrz) [reply app/controllers/concerns/ignore_not_found.rb]: This [issue](https://github.com/department-of-veterans-affairs/va.gov-team/issues/912) is in the backlog for updating how we tag services. Let's talk about it when the tools team is reformed.

Comment (kfrz) [reply app/controllers/concerns/ignore_not_found.rb]: We could do the exclusion configuration like so, which would look like:

```rb
module IgnoreNotFound
  def skip_sentry_exception_types
    Raven.configure { |c| c.excluded_exceptions += 'Common::Exceptions::RecordNotFound' }
  end
end
```

This to me is a bit less readable, and intertwines the logic for " what errors don't we want to log to APM " logic into our " which APM system are we using " logic, whereas if we were to swap to a different service (i.e. DataDog), we'd have to move this logic to a different configuration. Alternatively, we could do something more abstract:

```rb
module IgnoreNotFound
  def skip_sentry_exception_types
    ErrorLogging::SKIP_EXCEPTION_TYPES += [Common::Exceptions::RecordNotFound]
  end
end
```

That seems a bit further out of scope, and addresses a problem we don't currently have.

@johnpaulashenfelter @jholton @LindseySaari thoughts?

**Comment by jholton** [comment]
@kfrz Can you fill out the PR description please?

File: app/controllers/application_controller.rb [hunk 4]
```diff
@@ -61,6 +61,57 @@ def set_app_info_headers
...
+
+  def render_error(exc)
+    render json: { errors: exc.errors }, status: exc.status
+  end
+
+  # @note Later rescue_from statements take precedence and so we must take care to respect order
+  rescue_from StandardError do |error|
+    exc = Common::Exceptions::InternalServerError.new(error)
+    handle_error(exc)
+    render json: { errors: exc.errors }, status: exc.status_code
```
Comment (LindseySaari) [inline app/controllers/application_controller.rb]: Could we extend the render_error method to include json rendering for the exc.status_code case (lines 76 and 82 of this PR)?

File: spec/lib/common/client/error_handler_spec.rb [hunk 5]
```diff
@@ -16,69 +15,74 @@
...
     end
   end
 
   describe '#transformed_error' do
-  
     context 'when given an error without a specific counterpart' do
       subject { described_class.new(StandardError.new).transformed_error }
-      it 'returns the correct default error' do
+
+      xit 'returns the correct default error' do
```
Comment (LindseySaari) [inline spec/lib/common/client/error_handler_spec.rb]: Ah, these tests might not be applicable any longer after the rescue block breakout in the application controller

**Comment by omgitsbillryan** [comment]
Are you still looking for reviews on this PR as is? This PR is quite large and seems stale.

**Comment by kfrz** [comment]
@omgitsbillryan we'll likely re-request reviews Monday or Tuesday, after updating w/ `master` and ensuring build is green. We can try to break it out, it's a bit large and Danger is raising issues

File: app/controllers/application_controller.rb [hunk 6]
```diff
@@ -39,116 +39,81 @@ def routing_error
...
 
-  def user_context
-    {
-      uuid: current_user&.uuid,
-      authn_context: current_user&.authn_context,
-      loa: current_user&.loa,
-      mhv_icn: current_user&.mhv_icn
-    }
+  def handle_error(error)
+    Common::Client::ErrorHandler.handle(error)
```
Comment (kfrz) [inline app/controllers/application_controller.rb]: `Common::Client::ErrorHandler` would indicate the errors are originating in a client, though that's a separate subset of errors

File: app/controllers/application_controller.rb [hunk 7]
```diff
@@ -39,116 +39,81 @@ def routing_error
...
-        tags[:sign_in_method] = current_user.identity.sign_in[:service_name]
-        # account_type is filtered by sentry, becasue in other contexts it refers to a bank account type
-        tags[:sign_in_acct_type] = current_user.identity.sign_in[:account_type]
-      else
-        tags[:sign_in_method] = 'not-signed-in'
-      end
-    end
+  def render_error(exc)
+    render json: { errors: exc.errors }, status: exc.status
   end
```
Comment (kfrz) [inline app/controllers/application_controller.rb]: Signature here isn't great - use "exception" or something more descriptive than `exc` and perhaps add some comments

File: app/controllers/application_controller.rb [hunk 8]
```diff
@@ -39,116 +39,81 @@ def routing_error
...
+    handle_error(exc)
+    render_error(exc)
+  end
+
+  rescue_from Pundit::NotAuthorizedError do |error|
+    exc = Common::Exceptions::Forbidden.new(cause: error, detail: 'User does not have access to the requested resource')
+    handle_error(exc)
+    render_error(exc)
+  end
+
```
Comment (kfrz) [inline app/controllers/application_controller.rb]: In this case, we're actually reorganizing the control flow such that instead of rescuing *all* exceptions, including `ruby` exceptions, we're now rescuing from specific cases, and any `StandardErrors` will be converted into an `InternalServerError`.

Also on line 79, explain why `exc` instead of `error` is used

File: app/controllers/concerns/ignore_not_found.rb [hunk 9]
```diff
@@ -2,6 +2,6 @@
 
 module IgnoreNotFound
   def skip_sentry_exception_types
-    ApplicationController::SKIP_SENTRY_EXCEPTION_TYPES + [Common::Exceptions::RecordNotFound]
+    Raven.configure { |c| c.excluded_exceptions += 'Common::Exceptions::RecordNotFound' }
```
Comment (kfrz) [inline app/controllers/concerns/ignore_not_found.rb]: This could be a separate pr, including the error abstraction work

File: app/controllers/v0/health_care_applications_controller.rb [hunk 10]
```diff
@@ -5,6 +5,8 @@ class HealthCareApplicationsController < ApplicationController
     FORM_ID = '1010ez'
 
     skip_before_action(:authenticate)
+    before_action(:append_to_skip_sentry_exception_types)
+    after_action(:remove_from_skip_sentry_exception_types)
```
Comment (kfrz) [inline app/controllers/v0/health_care_applications_controller.rb]: I'm not sure this is a great pattern, it would likely be overlooked and cause headaches for devs. Let's rethink our approach here.

File: config/initializers/raven.rb [hunk 11]
```diff
@@ -13,4 +13,13 @@
...
   }
+
+  # appends to { Raven::Configuration::IGNORE_DEFAULT }
+  # @see https://docs.sentry.io/clients/ruby/config/
+  config.excluded_exceptions += %w[
+    Common::Exceptions::Unauthorized
+    Common::Exceptions::RoutingError
+    Common::Exceptions::Forbidden
+    Breakers::OutageException
+  ]
```
Comment (kfrz) [inline config/initializers/raven.rb]: Try to extract to separate PR

File: lib/common/client/error_handler.rb [hunk 12]
```diff
@@ -0,0 +1,15 @@
+# frozen_string_literal: true
+
+module Common
+  module Client
+    # This class is responsible for providing a public interface with which to perform common error handling tasks.
```
Comment (kfrz) [inline lib/common/client/error_handler.rb]: Rename to `Common::Error::Handler` or something not tied to "client"

File: lib/common/client/errors.rb [hunk 13]
```diff
@@ -10,22 +10,21 @@ class ClientError < Error
         attr_accessor :status
         attr_accessor :body
 
-        def initialize(message = nil, status = nil, body = nil)
+        def initialize(error = nil, message: nil, status: nil, body: nil)
```
Comment (kfrz) [inline lib/common/client/errors.rb]: Add documentation to this class

Also add/update docs in `va.gov-team` repo/README.md

File: lib/common/client/errors.rb [hunk 14]
```diff
@@ -10,22 +10,21 @@ class ClientError < Error
...
           super(message)
+          @cause = error
```
Comment (kfrz) [inline lib/common/client/errors.rb]: This class inherits from `StandardError`, so why do we have to *add* the cause attribute? 

It should be [built-in](https://ruby-doc.org/core-2.6/Exception.html) - add specs/documentation/comments to explain why this is necessary.

File: lib/common/exceptions/internal/forbidden.rb [hunk 15]
```diff
@@ -5,6 +5,7 @@ module Exceptions
     # Forbidden - We may eventually want different variations on this with distinct MinorCodes
     class Forbidden < BaseError
       def initialize(options = {})
+        @cause  = options[:cause]
```
Comment (kfrz) [inline lib/common/exceptions/internal/forbidden.rb]: This should be on the base error class

File: spec/lib/common/client/error_handler_spec.rb [hunk 16]
```diff
@@ -0,0 +1,88 @@
+# frozen_string_literal: true
+
+require 'rails_helper'
+
+describe Common::Client::ErrorHandler do
+  let(:gateway_timeout_error) { StandardError.new }
```
Comment (kfrz) [inline spec/lib/common/client/error_handler_spec.rb]: a-z

File: spec/lib/common/client/error_handler_spec.rb [hunk 17]
```diff
@@ -0,0 +1,88 @@
...
+  let(:unknown_format_error) { ActionController::UnknownFormat }
+  let(:not_safe_host_error) { Common::Exceptions::NotASafeHostError }
+  let(:base_error) { Common::Exceptions::BaseError }
+  let(:breakers_outage_error) { Breakers::OutageException }
+  let(:client_error) { Common::Client::Errors::ClientError }
+
+  describe '#log_error' do
+    subject { described_class.new(client_error.new) }
+
+    xit 'calls instance log error method' do
```
Comment (kfrz) [inline spec/lib/common/client/error_handler_spec.rb]: Don't skip me!

File: spec/lib/saml/settings_service_spec.rb [hunk 18]
```diff
@@ -2,7 +2,7 @@
 
 require 'rails_helper'
 require 'saml/settings_service'
-require 'lib/sentry_logging_spec_helper'
+require 'lib/shared_examples_for_sentry_logging'
```
Comment (kfrz) [inline spec/lib/saml/settings_service_spec.rb]: Easy to extract to separate PR

File: spec/request/health_care_applications_request_spec.rb [hunk 19]
```diff
@@ -296,6 +296,12 @@
 
           it 'renders error message' do
             expect(Raven).to receive(:capture_exception).with(error, level: 'error').once
+            # This test is failing because app/models/health_care_application.rb:38 is raising a
```
Comment (kfrz) [inline spec/request/health_care_applications_request_spec.rb]: :crying_cat_face:

File: spec/request/user_request_spec.rb [hunk 20]
```diff
@@ -181,6 +181,7 @@
     end
 
     it 'MVI success should only make a request to MVI one time per multiple requests!' do
+      pending 'after error handling changes this is calling MVI session class methods multiple times'
```
Comment (kfrz) [inline spec/request/user_request_spec.rb]: :crying_cat_face:

File: spec/spec_helper.rb [hunk 21]
```diff
@@ -23,6 +23,7 @@
     track_files '**/{app,lib}/**/*.rb'
 
     add_filter 'app/controllers/concerns/accountable.rb'
+    add_filter 'app/controllers/concerns/ignore_not_found.rb' # @todo remove this before merging
```
Comment (kfrz) [inline spec/spec_helper.rb]: Do this

**Comment by kfrz** [comment]
Closing this PR as it's complex and has been stale for awhile. 

Will backlog tasks and re-prioritize in the future.
